src/diffusion_for_multi_scale_molecular_dynamics/generators/langevin_generator.py

```diff
@@ -214,11 +228,97 @@ def atom_types_update(
         return a_im1
 
+    def adjust_atom_types_probabilities_for_greedy_sampling(
+        self,
+        one_step_transition_probs: torch.Tensor,
+        atom_types_i: torch.LongTensor,
+        u: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Update the transition probabilities and the gumbel random variables to allow greedy sampling.
+
+        At time step i, for every atom in a sample, we sample a random number. If it is larger than the probability of
+        that atom being in the MASK class, then we will sample greedily a new atom type (i.e. the most likely). To do
+        that, we simply replace the probability of the MASK class to zero and the gumbel noise u to zero. For non-MASK
+        atoms, we do nothing. For samples with only MASK atoms, we also do nothing.
+
+        Args:
+            one_step_transition_probs: class distributions at time t-1 given distribution at time t. p(a_{t-1} | a_t)
+            atom_types_i: indices of atom types at time i. Dimension: [number_of_samples, number_of_atoms]
+            u: gumbel noise used for sampling. Dimension: [number_of_samples, number_of_atoms, num_classes]
+
+        Returns:
+            one_step_transition_probs: probabilities are updated so a MASK to non-MASK transition can happen
+            u: set to a constant for samples with at least 1 non-MASK atom
+        """
+        # check which samples have at least 1 non-MASK atom
+        all_masked = torch.all(
+            atom_types_i == self.num_classes - 1, dim=-1
+        )  # dim: number_of_samples,
+
+        # we can only do greedy sampling for atoms that are masked
+        atom_is_masked = atom_types_i == self.num_classes - 1
+
+        # we will first erase the probability of staying MASK for some atoms randomly by drawing from a binary
+        # distribution given by one_step_transition_probs[:, :, -1] i.e. the probabilities related to the MASK class.
+        # sample to override the MASK probability as the most likely
+        binary_sample = self._draw_binary_sample(atom_types_i.shape[0]).to(
+            device=atom_types_i.device
+        )
+        unmask_this_sample = binary_sample > one_step_transition_probs[:, :, -1]
+        # if we override the MASK probability & there's already a non-MASK sample & that atom is masked,
+        # use a greedy sampling for that atom
+        do_greedy_sampling = torch.logical_and(
+            ~all_masked.view(-1, 1),
+            unmask_this_sample,
+        )
+        do_greedy_sampling = torch.logical_and(do_greedy_sampling, atom_is_masked)
+        # replace the probability of getting a mask for those by 0 - so that stat cannot be sampled
+        one_step_transition_probs[:, :, -1] = torch.where(
+            do_greedy_sampling, 0, one_step_transition_probs[:, :, -1]
+        )
+
+        # replace u with a constant for samples with a non-MASK token present - this ensures a greedy sampling
```

Comment:
As we discussed, it might be useful to add a comment that this essentially means predicting MASK given our beta schedule...